service/storage_service.cc

```diff
@@ -435,36 +422,41 @@ future<> storage_service::topology_state_load() {
 
         tmptr->set_version(_topology_state_machine._topology.version);
 
-        auto update_topology = [&] (locator::host_id id, inet_address ip, const replica_state& rs) {
+        auto update_topology = [&] (locator::host_id id, std::optional<inet_address> ip, const replica_state& rs) {
             tmptr->update_topology(id, locator::endpoint_dc_rack{rs.datacenter, rs.rack},
                                    to_topology_node_state(rs.state), rs.shard_count);
-            tmptr->update_host_id(id, ip);
+            if (ip) {
+                tmptr->update_host_id(id, *ip);
+            }
         };
 
         auto add_normal_node = [&] (raft::server_id id, const replica_state& rs) -> future<> {
             locator::host_id host_id{id.uuid()};
-            auto ip = co_await id2ip(id);
+            auto ip = am.find(id);
 
             slogger.trace("raft topology: loading topology: raft id={} ip={} node state={} dc={} rack={} tokens state={} tokens={} shards={}",
                           id, ip, rs.state, rs.datacenter, rs.rack, _topology_state_machine._topology.tstate, rs.ring.value().tokens, rs.shard_count);
             // Save tokens, not needed for raft topology management, but needed by legacy
             // Also ip -> id mapping is needed for address map recreation on reboot
-            if (!is_me(ip)) {
+            if (is_me(host_id)) {
+                co_await _sys_ks.local().update_tokens(rs.ring.value().tokens);
+                co_await _gossiper.add_local_application_state({{ gms::application_state::STATUS, gms::versioned_value::normal(rs.ring.value().tokens) }});
+            } else if (ip && !is_me(*ip)) {
+                // In replace-with-same-ip scenario the replaced node IP will be the same
+                // as ours, we shouldn't put it into system.peers.
```

Comment:
If this is indeed an old Raft log entry that we're applying, which comes from a time where a replace-with-same-IP was happening, then this check might not always detect the case -- if `am.find` fails, right?

Maybe there's a more reliable way to detect the replace-with-same-IP scenario, from topology state or sth.

Reply:
> If am.find fails, we won't pass the condition on ip being present in this if.

That's my point and my question is whether it's a problem. We won't enter this `else if` branch, meaning we won't detect the replace-with-same-ip scenario, meaning we will execute the wrong logic for replace-with-same-ip.

Every time we enter this `else if`, we are guaranteed to be in replace-with-same-ip. But not every time we're in replace-with-same-ip, we will enter this `else if` -- because `am.find` is unreliable (it depends on timing). So we will not call `co_await _sys_ks.local().update_peer_info(*ip, info)` sometimes during replace-with-same-ip.

Reply:
We can only be the `replacing` node, the `being_replaced` node with this IP should be stopped and its state is not important.

The condition `if (is_me(host_id))` being false means we are considering some other normal node. The missing ip means we are postponing the local system table update until later times. This can be resolved either by topology coordinator, when it calls `wait_for_ip` for the actual new nodes (not some old irrelevant ones), of by `gossiper_state_change_subscriber_proxy` in case the `wait_for_ip` failed.

If the ip is known and it's equal to our IP, then we don't need to update `system.peers`. We should wait until the `being_replaced` become left, and then it will be removed from system.peers by the relevant code in `process_left_node`.